src/common/function/src/scalars/sleep.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::time::Duration;
+use std::{fmt, thread};
+
+use common_query::error::{InvalidFuncArgsSnafu, Result};
+use common_query::prelude::{Signature, Volatility};
+use datatypes::prelude::ConcreteDataType;
+use datatypes::value::Value;
+use datatypes::vectors::{Int64Vector, VectorRef};
+use snafu::ensure;
+
+use crate::function::{Function, FunctionContext};
+use crate::function_registry::FunctionRegistry;
+
+/// Sleep function that pauses execution for specified seconds
+#[derive(Clone, Debug, Default)]
+pub(crate) struct SleepFunction;
+
+impl SleepFunction {
+    pub fn register(registry: &FunctionRegistry) {
+        registry.register(Arc::new(SleepFunction));
+    }
+}
+
+const NAME: &str = "sleep";
+
+impl Function for SleepFunction {
+    fn name(&self) -> &str {
+        NAME
+    }
+
+    fn return_type(&self, _input_types: &[ConcreteDataType]) -> Result<ConcreteDataType> {
+        Ok(ConcreteDataType::int64_datatype())
+    }
+
+    fn signature(&self) -> Signature {
+        // Accept int32, int64 and float64 types
+        Signature::uniform(
+            1,
+            vec![
+                ConcreteDataType::int32_datatype(),
+                ConcreteDataType::int64_datatype(),
+                ConcreteDataType::float64_datatype(),
+            ],
+            Volatility::Volatile,
+        )
+    }
+
+    fn eval(&self, _func_ctx: FunctionContext, columns: &[VectorRef]) -> Result<VectorRef> {
+        ensure!(
+            columns.len() == 1,
+            InvalidFuncArgsSnafu {
+                err_msg: format!(
+                    "The length of the args is not correct, expect exactly one, have: {}",
+                    columns.len()
+                ),
+            }
+        );
+
+        let vector = &columns[0];
+        let mut result = Vec::with_capacity(vector.len());
+
+        for i in 0..vector.len() {
+            let secs = match vector.get(i) {
+                Value::Int64(x) => x as f64,
+                Value::Int32(x) => x as f64,
+                Value::Float64(x) => x.into_inner(),
+                _ => {
+                    result.push(None);
+                    continue;
+                }
+            };
+            // Sleep for the specified seconds TODO: use tokio::time::sleep when the scalars are async
+            thread::sleep(Duration::from_secs_f64(secs));
```

Comment:
I'd suggest not to sleep in a scalar function. This pauses the runtime thread which is risky.